Native S2I Builder image

vorburger opened the pull request.

vorburger committed:
add Native S2I Builder image v0.1 (#304)

vorburger committed:
add 'EXPOSE 8080' to native S2I builder Dockerfile

vorburger commented:
How cool is this? (for #304)

@siamaksade @cescoffier @dmlloyd @gsmet @geoand @gastaldi

geoand commented:
Looks great!

gastaldi commented on docker/centos-graal-maven-s2i/Dockerfile:
```diff
@@ -0,0 +1,57 @@
+ARG MAVEN_VERSION=3.6.0
+
+ENV GRAAL_CE_URL=https://github.com/oracle/graal/releases/download/vm-${GRAAL_VERSION}/graalvm-ce-${GRAAL_VERSION}-linux-amd64.tar.gz
+ENV MAVEN_URL=http://apache.osuosl.org/maven/maven-3/$MAVEN_VERSION/binaries/apache-maven-$MAVEN_VERSION-bin.tar.gz
```
```suggestion
ENV MAVEN_URL=https://archive.apache.org/dist/maven/maven-3/$MAVEN_VERSION/binaries/apache-maven-$MAVEN_VERSION-bin.tar.gz
```

gastaldi commented on docker/centos-graal-maven-s2i/Dockerfile:
```diff
@@ -0,0 +1,57 @@
+LABEL io.openshift.s2i.scripts-url="image:///usr/local/s2i" \
+      io.openshift.s2i.destination="/tmp" \
+
+      io.k8s.description="Platform for building and running Quarkus.io NATIVE Java applications" \
```
```suggestion
      io.k8s.description="Platform for building and running Quarkus.io Native Java applications" \
```

gastaldi commented on docker/centos-graal-maven-s2i/Dockerfile:
```diff
@@ -0,0 +1,57 @@
+      io.openshift.s2i.destination="/tmp" \
+
+      io.k8s.description="Platform for building and running Quarkus.io NATIVE Java applications" \
+      io.k8s.display-name="Quarkus.io NATIVE Java Applications" \
```
```suggestion
      io.k8s.display-name="Quarkus.io Native Java Applications" \
```

gastaldi reviewed (approved):
LGTM. Added some minor suggestions, feel free to approve or reject them :smile:

gastaldi committed:
Minor label spelling update docker/centos-graal-maven-s2i/Dockerfile

Co-Authored-By: vorburger <mike@vorburger.ch>

gastaldi committed:
use better Maven DL URL in Native S2I  Dockerfile

Co-Authored-By: vorburger <mike@vorburger.ch>

gastaldi reviewed (approved):
LGTM

geoand commented:
Very good stuff!

cescoffier commented on docker/centos-graal-maven-s2i/Dockerfile:
```diff
@@ -0,0 +1,57 @@
+
+ARG GRAAL_VERSION
+ENV GRAAL_VERSION=${GRAAL_VERSION:-1.0.0-rc12}
+ARG MAVEN_VERSION=3.6.0
```
Wondering if we can reuse one of the docker image Quarkus provides: https://github.com/quarkusio/quarkus/tree/master/docker

Avoiding having to maintain version is several places would be great.

vorburger replied:
@cescoffier saw those and took inspiration from them.. the problem with doing e.g. a `FROM centos-graal-native-image` is that it makes it a bit harder to then set up CI/CD of this container - systems like Quay and Docker Hub except to run (exactly) 1 Dockerfile... so, at least for a first iteration, it seems simpler to me to keep it all together.

cescoffier replied:
Ok

cescoffier commented on docker/centos-graal-maven-s2i/Dockerfile:
```diff
@@ -0,0 +1,57 @@
+LABEL io.openshift.s2i.scripts-url="image:///usr/local/s2i" \
+      io.openshift.s2i.destination="/tmp" \
+
+      io.k8s.description="Platform for building and running Quarkus.io Native Java applications" \
```
The wording is not totally in line with the website - check https://github.com/quarkusio/quarkus/blob/master/docs/src/main/asciidoc/0-glossary.adoc.

vorburger replied:
I've attempted to adhere to that in an upcoming update (and raised #1377), hopefully better now.

cescoffier commented on docker/centos-graal-maven-s2i/README.md:
```diff
@@ -0,0 +1,30 @@
+
+### OpenShift
+
+    oc new-build https://github.com/quarkusio/quarkus.git --context-dir=docker/centos-graal-maven-s2i --name quarkus-native-s2i
```
Nice, just learned something :-D

cescoffier commented on docker/centos-graal-maven-s2i/README.md:
```diff
@@ -0,0 +1,30 @@
+
+### OpenShift
+
+    oc new-app quarkus-native-s2i~https://github.com/quarkusio/quarkus-quickstarts --context-dir=getting-started-native --name=getting-started-native
```
Should we remind the user to have "installed" with the command mentioned previously?

vorburger replied:
I'll re-shuffle the documents sections to make more sense, see if you like it better after the next update.

cescoffier commented on docker/centos-graal-maven-s2i/README.md:
```diff
@@ -0,0 +1,30 @@
+    oc new-app quarkus-native-s2i~https://github.com/quarkusio/quarkus-quickstarts --context-dir=getting-started-native --name=getting-started-native
+
+Note that GraalVM-based native build are more memory & CPU intensive than regular pure Java builds,
+and you therefore may need to increase the quota for OpenShift's S2I builder containers.
```
Can you give some hint about this? 
I guess you can edit the build object.

vorburger replied:
I was hoping someone else would help with how exactly to do that... :smiley:

vorburger replied:
There now is at least some doc. Could always be further improved, but god enough for a 0.1?

cescoffier commented on docker/centos-graal-maven-s2i/s2i/assemble:
That's a lot of skip :-) But I'm all good with that.

cescoffier replied:
Maybe you can use multiline as the `-e -B` are quite important (and hidden right now)

vorburger replied:
FYI The skip are [are stolen from here](https://github.com/fabric8io-images/s2i/blob/master/java/templates/s2i/assemble#L123). I've moved `-e -B` to the front and made it multi line.

cescoffier commented on docker/centos-graal-maven-s2i/s2i/run:
For the record, there are discussions about the 0.0.0.0 thingy.

vorburger replied:
Yup, so I gathered.. if/when that happens, someone needs to drop this from a whole lot of quickstarts and docs and scripts and all - including this. For now, it's required though.

cescoffier reviewed (changes requested):
That's pretty cool, I've made a few comments.

cescoffier commented:
BTW, how will we version the S2I, during the release process? If so how does that work?

vorburger committed:
improve graal-maven-s2i Dockerfile io.k8s.* LABELs

as suggested by @cescoffier during code review

vorburger committed:
move -e -B in GraalVM S2I assemble script (and multiline)

vorburger committed:
restructure GraalVM S2I README sections (clearer like this)

vorburger commented:
> BTW, how will we version the S2I, during the release process? If so how does that work?

Not 100% sure I know what you mean, but if you are asking how we will release a container image from this Dockerfile to e.g. Quay.io (and/or Docker Hub), then that's trivial to set up (IFF you keep a single `Dockerfile` ... see above), and it can easily be made to "tag" such images based on the branch in this repo. Does this help?

cescoffier commented on docker/centos-graal-maven-s2i/Dockerfile:
```diff
@@ -0,0 +1,57 @@
+FROM centos:latest
+
+ARG GRAAL_VERSION
+ENV GRAAL_VERSION=${GRAAL_VERSION:-1.0.0-rc12}
```
Need to be updated to 13.

vorburger replied:
done! Thanks for noticing.

cescoffier commented on docker/centos-graal-maven-s2i/Dockerfile:
```diff
@@ -0,0 +1,57 @@
+# Original Author: Michael Vorburger.ch
+
+# This part is heavily inspired by ../centos-graal/Dockerfile
+FROM centos:latest
```
Do we really want to use "latest"?

vorburger replied:
Using e.g. `centos:7` is actually meaningless and doesn't work the way you think it does, as far as I understand... I learnt this the hard way over here https://github.com/fabric8io-images/s2i/issues/206#issuecomment-445047444 and here https://github.com/fabric8io-images/s2i/issues/172 full background logged here https://gist.github.com/vorburger/ea367a76f69aafb8260c62f6bedf2788... Unless all that's only for 7.5/7.6 and not 8, but I doubt it. I am still happy to change it to `centos:7` anyway if you like, but doubt that's really useful and does what most expect.

cescoffier commented on docker/centos-graal-maven-s2i/Dockerfile:
```diff
@@ -0,0 +1,57 @@
+
+ENV PATH=$PATH:"/usr/local/s2i"
+
+RUN mkdir /quarkus.io/
```
Really minor, but I try to about `.` in my directory name. What about just "quarkus"

vorburger replied:
it's now `/home/quarkus` OK for you?

cescoffier replied:
Yes

cescoffier commented on docker/centos-graal-maven-s2i/Dockerfile:
```diff
@@ -0,0 +1,57 @@
+
+RUN mkdir /quarkus.io/
+
+EXPOSE 8080
```
Any reason to expose the port?

vorburger replied:
Just makes it easier to use `oc new-app` / `oc expose`  (and `docker run`) without having to re-specify `8080` ... of course one could change it, but ... people probably typically will not? And even if they did (and therefore specified another port in `oc new-app` / `oc expose`  (and `docker run`), this doesn't cause any "harm" either, for the convenience it offers? (See also https://github.com/fabric8io-images/s2i/issues/115 and https://github.com/fabric8io-images/s2i/pull/205 for some history.)

cescoffier replied:
Thanks for the explanation.

cescoffier reviewed (changes requested):
A few comments

vorburger committed:
minor centos-graal-maven-s2i README update

vorburger committed:
add 'USER 1001' in centos-graal-maven-s2i/Dockerfile

This should fix this S2I build error seen on OpenShift:

build error: image "..." must specify a user that is numeric and within the range of allowed users

vorburger committed:
use 'cp -v' in s2i/assemble

vorburger committed:
improve centos-graal-maven-s2i README

vorburger committed:
Merge branch 'master' into issue-304_s2i-native-DEV

* master: (167 commits)
  Allow for usage of @ConfigProperty on fields without the need for @Inject. Fixes #1315.
  Expose all flavors (core, axle, rx) of Vert.x EventBus
  RestClient - passing incoming JAX-RS headers, fixes #1017
  JAX-RS body writers for Vert.x JsonObject and JsonArray
  Increase CI memory
  PathTestHelper should accept default Gradle build paths now + make sure we run the test in gradle-it
  Change to use ./mvnw
  Update opentracing guide to describe using application.properties as alternative to env vars
  fix npe in dev mode when source is missing in recompilation diagnostic
  Add the quarkus-elytron-security component in the BOM
  Moves the logic that handles the registration of schema class for serialization in smallrye-openapi-extension
  Add Maven Wrapper to generated projects
  [security extension] - AuthConfig only configured if IdentityManager is produced
  #1522 Warning to keep H2 version as is
  Minor corrections to mx and git repo clone instructions.
  programmatic client generation for MP Rest Client, fixes #1317
  Hibernate ORM - Fix config property name in error message
  Camel - Use a constant for Camel root package directory
  Micro version updates
  JAX-RS Generate reflection declaration from MP OpenAPI schema definition
  ...

vorburger committed:
bump GRAAL_VERSION rc12 to rc13

vorburger committed:
use ARG instead of ENV in Dockerfile

vorburger committed:
move add user up, and use home dir in MAVEN_CONFIG

vorburger committed:
use /home/quarkus instead of /quarkus.io

vorburger committed:
use ENV instead of ARG in Dockerfile where appropriate

vorburger commented:
@cescoffier I've further tested this, and (from my side) this is good for merge now.

cescoffier commented:
Waiting for CI and will merge.

geoand commented on docker/centos-graal-maven-s2i/s2i/run:
`-Dquarkus.http.host=0.0.0.0` shouldn't be needed anymore since it's now the default

vorburger replied:
But is that already released, right now? I would prefer to keep this and then see it removed a while after a new release, together with more [generally cleaning this up in many places where it currently still appears](https://github.com/search?q=org%3Aquarkusio+quarkus.http.host&type=Code).

geoand replied:
The fix is in master and will be present in the `0.12.0` release that I think @cescoffier will be doing tomorrow. 
I'll leave it up to you to best decide how to handle this :)

geoand replied:
https://github.com/quarkusio/quarkus/pull/1312#issuecomment-474271302

vorburger commented on docker/centos-graal-maven-s2i/Dockerfile:
```diff
@@ -0,0 +1,61 @@
+# This part is heavily inspired by ../centos-graal/Dockerfile
+FROM centos:latest
+
+ARG GRAAL_VERSION=1.0.0-rc13
```
Yup, just like in [a few other places where it appears](https://github.com/search?q=org%3Aquarkusio+1.0.0-rc13&type=Code).

geoand replied:
:+1

vorburger commented:
> Waiting for CI and will merge.

@cescoffier merge?

vorburger committed:
update Native S2I README with testing results re. CPU limit

vorburger committed:
add 'oc patch' tip to Native S2I README (tx @tqvarnst)

cescoffier commented:
Thanks!
